File: platforms/mesh_host/android/src/main/kotlin/com/frazko/mesh_host/BleWriteQueue.kt
package com.frazko.mesh_host

/** ATT writes/notifications own one fragment until their completion callback.
 * Enqueuing another record must never start a second write or drop the first. */
internal class BleWriteQueue {
    private val pending = ArrayDeque<ByteArray>()
    private var inFlight = false
    val idle: Boolean get() = !inFlight && pending.isEmpty()

    fun addAll(parts: List<ByteArray>) { pending.addAll(parts) }
    fun begin(): ByteArray? {
        if (inFlight || pending.isEmpty()) return null
        inFlight = true
        return pending.first()
    }
    fun complete() {
        if (!inFlight) return
        pending.removeFirst()
        inFlight = false
    }
    fun refused() { inFlight = false }
}

/** Durable retransmission belongs to a new transport session. Repeated product
 * sends on a healthy reliable link must not append the entire outbox again. */
internal class DurableRecordHistory {
    private val records = LinkedHashSet<String>()
    private val pending = ArrayDeque<ByteArray>()
    fun enqueue(payloads: List<ByteArray>) {
        // Prioritize newly submitted records before the unsent reconnect tail.
        // This is plaintext: Noise counters are assigned only at dequeue.
        val fresh = payloads.filter { admit(it) }
        fresh.asReversed().forEach { pending.addFirst(it) }
    }
    fun next(): ByteArray? = if (pending.isEmpty()) null else pending.removeFirst()
    fun admit(payload: ByteArray): Boolean {
        if (payload.size < 96 || payload[0] != 0x72.toByte() || payload[1] != 1.toByte()) return true
        val key = java.security.MessageDigest.getInstance("SHA-256")
            .digest(payload).joinToString("") { "%02x".format(it) }
        if (!records.add(key)) return false
        if (records.size > 8192) records.remove(records.first())
        return true
    }
}

File: platforms/mesh_host/ios/mesh_host/Sources/mesh_host/BleWriteQueue.swift
import Foundation
import CryptoKit

/// Keeps the head fragment until ATT completion (or notification acceptance).
/// New records cannot interrupt an in-flight write or reorder Noise counters.
final class BleWriteQueue {
  private var pending = [[UInt8]]()
  private var head = 0
  private var inFlight = false
  var idle: Bool { !inFlight && head == pending.count }

  func append(_ parts: [[UInt8]]) { pending.append(contentsOf: parts) }
  func begin() -> [UInt8]? {
    guard !inFlight, head < pending.count else { return nil }
    inFlight = true
    return pending[head]
  }
  func complete() {
    guard inFlight else { return }
    head += 1
    inFlight = false
    if head == pending.count { pending.removeAll(keepingCapacity: true); head = 0 }
  }
  func refused() { inFlight = false }
}

final class DurableRecordHistory {
  private var records = Set<Data>()
  private var order = [Data]()
  private var pending = [[UInt8]]()
  func enqueue(_ payloads: [[UInt8]]) {
    // Assign Noise counters only when removing the next record for the radio.
    pending.insert(contentsOf: payloads.filter { admit($0) }, at: 0)
  }
  func next() -> [UInt8]? { pending.isEmpty ? nil : pending.removeFirst() }
  func admit(_ payload: [UInt8]) -> Bool {
    guard payload.count >= 96, payload[0] == 0x72, payload[1] == 1 else { return true }
    let key = Data(SHA256.hash(data: Data(payload)))
    guard records.insert(key).inserted else { return false }
    order.append(key)
    if order.count > 8192 { records.remove(order.removeFirst()) }
    return true
  }
}
